osu.Game/Database/BackgroundDataStoreProcessor.cs

```diff
@@ -436,6 +437,95 @@ private void upgradeScoreRanks()
+            double OldMultiplier,
+            double NewMultiplier,
+            int ScoreVersion);
+
+        private static readonly ModMultiplierChange[] mod_multiplier_changes =
+        [
+            new ModMultiplierChange("CL", "taiko", 0.96, 1, 30000017),
+            new ModMultiplierChange("CL", "fruits", 0.96, 1, 30000017),
+            new ModMultiplierChange("CL", "mania", 0.96, 1, 30000017),
+        ];
+
+        private void upgradeModMultipliers()
+        {
+            Logger.Log("Performing mod multiplier upgrades...");
+
+            int latestMultiplierChange = mod_multiplier_changes.Max(change => change.ScoreVersion);
+
+            var scores = realmAccess
+                         .Run(r => r.All<ScoreInfo>()
+                                    .Where(score => score.TotalScoreVersion < latestMultiplierChange && !score.BackgroundReprocessingFailed)
+                                    .Detach())
+                         // must be done after materialisation, as realm doesn't support
+                         // filtering on nested property predicates or projection via `.Select()`
+                         .Where(s => s.Ruleset.IsLegacyRuleset())
+                         .ToList();
+
+            if (scores.Count == 0)
+                return;
+
+            var notification = showProgressNotification(scores.Count, "Adjusting mod multipliers for scores", "scores now use latest mod multipliers");
+
+            int processedCount = 0;
+            int failedCount = 0;
+
+            foreach (var score in scores)
+            {
+                if (notification?.State == ProgressNotificationState.Cancelled)
+                    break;
+
+                updateNotificationProgress(notification, processedCount, scores.Count);
+
+                sleepIfRequired();
+
+                double? newTotalScore = null;
+
+                try
+                {
+                    foreach (var multiplierChange in mod_multiplier_changes)
+                    {
+                        if (score.TotalScoreVersion >= multiplierChange.ScoreVersion
+                            || score.Ruleset.ShortName != multiplierChange.RulesetName
+                            || score.APIMods.All(m => m.Acronym != multiplierChange.ModAcronym))
+                        {
+                            continue;
+                        }
+
+                        newTotalScore ??= score.TotalScore;
+                        newTotalScore /= multiplierChange.OldMultiplier;
+                        newTotalScore *= multiplierChange.NewMultiplier;
+                    }
+
+                    realmAccess.Write(r =>
+                    {
+                        var refetched = r.Find<ScoreInfo>(score.ID)!;
+                        if (newTotalScore != null)
+                            refetched.TotalScore = (long)newTotalScore.Value;
+                        refetched.TotalScoreVersion = LegacyScoreEncoder.LATEST_VERSION;
```

Comment:
We're doing this in three different methods in this processor class now, alongside local queries for current version. This means if two migrations need to happen, only the first to execute will actually get run, unless I'm missing something.

May need to re-think how these are being applied. Either applying all total score changes in one method, or updating the `TotalScoreVersion` as a final step instead of locally?

Reply:
Another option would be changing the version value to be incremented once per change (similar to realm versioning) and specifically apply migrations based on the imminent version number (then incrementing by one on application), but that's a bit of an annoying structural/logic change to make initially.